### PR TITLE
chains: add QA and summarization using map reduce

### PR DESCRIPTION
**Pull request by FluffyKebab**
This fixes some problems with the input keys in the map reduce chain and adds QA and summarization using the map reduce document chain. The apply function is also updated such that outputs and inputs are at the same index. 

### PR Checklist

- [x] Read the [Contributing documentation](https://github.com/tmc/langchaingo/blob/main/CONTRIBUTING.md).
- [x] Read the [Code of conduct documentation](https://github.com/tmc/langchaingo/blob/main/CODE_OF_CONDUCT.md).
- [x] Name your Pull Request title clearly, concisely, and prefixed with the name of the primarily affected package you changed according to [Good commit messages](https://go.dev/doc/contribute#commit_messages) (such as `memory: add interfaces for X, Y` or `util: add whizzbang helpers`).
- [x] Check that there isn't already a PR that solves the problem the same way to avoid creating a duplicate.
- [x] Provide a description in this PR that addresses **what** the PR is solving, or reference the issue that it solves (e.g. `Fixes #123`).
- [x] Describes the source of new concepts.
- [x] References existing implementations as appropriate.
- [x] Contains test coverage for new functions.
- [x] Passes all [`golangci-lint`](https://golangci-lint.run/) checks.

### CODE REVIEW

File: chains/chains.go [hunk 1]
```diff
@@ -123,46 +124,82 @@ func Predict(ctx context.Context, c Chain, inputValues map[string]any, options .
...
-					result map[string]any
-					err    error
-				}{
-					result: res,
-					err:    err,
+			defer wg.Done()
+			for {
+				select {
+				case <-ctx.Done():
+					return
```
Comment (tobiade) [inline chains/chains.go:156]: We can still write to `resultsChan` here with the error returned by `ctx.Err` i.e.
```go
resultsChan <- applyResult{
    err: ctx.Err()
    // can add a sentinel value for `i` e.g. -1 if needed, but don't think it's that important atm
}
```
That enables the client to know if the `Apply` function failed due to a timeout/cancellation.

Comment (tobiade) [reply chains/chains.go:156]: Ignore - I've just seen `getApplyResults` 😄

File: chains/chains_test.go [hunk 2]
```diff
@@ -38,3 +41,24 @@ func TestApply(t *testing.T) {
 	require.NoError(t, err)
 	require.Equal(t, numInputs, len(results), "number of inputs and results not equal")
 }
+
+func TestApplyWithCanceledContext(t *testing.T) {
```
Comment (tobiade) [inline chains/chains_test.go:45]: We should uplift the `TestApply` function to be more comprehensive - it should test that an expected output exists for a given input.

File: chains/chains.go (hunk 1 above)
Comment (tmc) [reply chains/chains.go:156]: There’s also cancel reason in newer go releases.

File: chains/chains_test.go [hunk 3]
```diff
@@ -13,6 +15,7 @@ import (
 type testLanguageModel struct{}
 
 func (l testLanguageModel) GeneratePrompt(_ context.Context, _ []schema.PromptValue, _ ...llms.CallOption) (llms.LLMResult, error) { //nolint:lll
+	time.Sleep(time.Second)
```
Comment (tmc) [inline chains/chains_test.go:18]: intentional?

**Review by tmc — APPROVED** [review]
LGTM

**Review by tmc — APPROVED** [review]
LGTM

File: chains/chains_test.go (hunk 3 above)
Comment (FluffyKebab) [reply chains/chains_test.go:18]: It is not necessary. I can remove it.

**Pull request merged**